test/rules/unnecessary_new_test.dart

```diff
@@ -0,0 +1,136 @@
+// Copyright (c) 2023, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:test_reflective_loader/test_reflective_loader.dart';
+
+import '../rule_test_support.dart';
+
+main() {
+  defineReflectiveSuite(() {
+    defineReflectiveTests(UnnecessaryNewTest);
+  });
+}
+
+@reflectiveTest
+class UnnecessaryNewTest extends LintRuleTest {
+  @override
+  String get lintRule => 'unnecessary_new';
+
+  test_localFinal_new() async {
+    await assertDiagnostics(r'''
+class A {
+  const A();
+}
+
+void f() {
+  final a = new A();
```

Comment:
Not sure whether it's useful to test all the locations in which an instance creation expression can appear, so we might consider dropping `test_localFinal_new` and `test_localFinal_noNew`.

Reply:
I'm certain that it's not useful. :/ I'm just trying to mechanically migrate these to be reflective. But I guess I'm happy to _delete_ uninteresting cases.

Reply:
If we don't delete them now then I'm not sure we'll come back to delete them later. That said, I'm not asking for perfection, just pointing out the things I see. But you're right, it more closely reflects the original tests this way, so I'm fine leaving it as-is.